### PR TITLE
Add lawn measurements

### PR DESCRIPTION
**Pull request by drew-sinha**
This pull request provides automated lawn finding with sklearn's Gaussian mixture modeling (GMM). At the core, a 2-component GMM is fit to image histogram of a single image and a threshold for lawn_segmentation is computed as the classification boundary between the two Gaussians. An annotator (annotate_lawn) constructs a lawn by combining the lawns found from several of the first images in a position timecourse (as a bitwise or) and saves the lawn to disk (+ adds the lawn area to the position annotations).  A measurement class also makes measurements of background and lawn intensities using the automatically generated mask.

There are a couple of things that could use feedback (also commented below....):

1. Is it worth it to add the experiment_metadata (or general position annotations) to measurement class signature? Here, I load it for the optocoupler & vignette mask. In the workflow of using the metadata more (+ the automated vignette masks from process_images in this instance), it seems like it might make sense.

2. Is there a well-constrained way to include lawn mask data in the position annotations? Right now, one of the two major bottlenecks is measurement of lawn parameters. Naively, I can only imagine that this is due to reading from disk. This happens twice in the measurements - once for the image (can't help that), and again for the position's lawn mask.

### CODE REVIEW

File: elegant/process_data.py [hunk 1]
```diff
@@ -80,6 +85,75 @@ def annotate_stage_pos(experiment_root, position, metadata, annotations):
...
+    microns_per_pixel = process_images.microns_per_pixel(metadata['objective'],metadata['optocoupler'])
+
+    position_images = load_data.scan_experiment_dir(experiment_root)[position]
+    first_imagepaths = []
+    for timepoint, timepoint_images in position_images.items():
+        if position_root / (timepoint + ' bf.png') in timepoint_images:
+            first_imagepaths.append(str(position_root / (timepoint + ' bf.png')))
+        if len(first_imagepaths) == num_images_for_lawn:
+            break
+
```
Comment (zpincus) [inline elegant/process_data.py]: Above stanza seems over-complicated. Why not something like:
```python
sorted(position_root.glob('* bf.png'))[:num_images_for_lawn]
```
though perhaps I'm missing a bit of subtlety to what that if-statement at line 101 is doing? (I.e. when would that statement not be true??)

File: elegant/process_data.py [hunk 2]
```diff
@@ -80,6 +85,75 @@ def annotate_stage_pos(experiment_root, position, metadata, annotations):
...
+    first_images = list(map(freeimage.read, first_imagepaths))
```
Comment (zpincus) [inline elegant/process_data.py]: Generally I prefer list comprehensions to `list(map(...))` just because it's a bit more readable. Especially as you're doing comprehensions below anyway.

File: elegant/process_data.py [hunk 3]
```diff
@@ -80,6 +85,75 @@ def annotate_stage_pos(experiment_root, position, metadata, annotations):
...
+    first_images = [process_images.pin_image_mode(image, optocoupler=metadata['optocoupler'])
+        for image in first_images]
+
+    individual_lawns = [gmm_lawn_maker(image, metadata['optocoupler']) for image in first_images]
+    lawn_mask = numpy.bitwise_or.reduce(individual_lawns, axis=0)
+
+    freeimage.write(lawn_mask.astype('uint8')*255, str(lawn_mask_root / f'{position}.png')) # Some better way to store this mask in the annotations?
```
Comment (zpincus) [inline elegant/process_data.py]: Better to store in an image file, unless you want to store it as a polygon or something in the annotation file. But no reason to keep raster data in anything other than an image file. You could use [skimage.measure.find_contours](http://scikit-image.org/docs/dev/auto_examples/edges/plot_contours.html) to get a polygon from the mask. But mask-files are probably easier and simpler for others to use...

File: elegant/process_data.py [hunk 4]
```diff
@@ -80,6 +85,75 @@ def annotate_stage_pos(experiment_root, position, metadata, annotations):
...
+    annotations['lawn_area'] = lawn_mask.sum() * microns_per_pixel**2
+
+def gmm_lawn_maker(image, optocoupler, return_model=False):
```
Comment (zpincus) [inline elegant/process_data.py]: I'd say move this (and related imports) to `elegant.segment_images` to keep this file cleaner.

File: elegant/process_data.py [hunk 5]
```diff
@@ -478,3 +552,35 @@ def get_mask(self, position_root, derived_root, timepoint, annotations):
...
         else:
             return freeimage.read(mask_file)
+
+class LawnMeasurements:
+    feature_names = ['summed_lawn_intensity', 'median_lawn_intensity', 'background_intensity']
+
+    def measure(self, position_root, timepoint, annotations, before, after):
+        measures = {}
+
+        # Load metadata.... TODO ask Zach if this can be a feature of the measurement signature
```
Comment (zpincus) [inline elegant/process_data.py]: This is a good idea. Will you do a separate PR to add that to the measurements?

File: elegant/process_data.py [hunk 6]
```diff
@@ -478,3 +552,35 @@ def get_mask(self, position_root, derived_root, timepoint, annotations):
...
+        timepoint_imagepath = position_root / (timepoint + ' bf.png')
+        timepoint_image = freeimage.read(timepoint_imagepath)
+        rescaled_image = process_images.pin_image_mode(timepoint_image, optocoupler=metadata['optocoupler'])
+
+        lawn_mask = freeimage.read(experiment_root / 'derived_data' / 'lawn_masks' / f'{position_name}.png').astype('bool')
+        vignette_mask = process_images.vignette_mask(metadata['optocoupler'], timepoint_image.shape)
+
+        # Remove the animal from the lawn if possible.
+        center_tck, width_tck = annotations.get('pose', (None, None))
+        if center_tck is None:
```
Comment (zpincus) [inline elegant/process_data.py]: simpler:
```python
if center_tck is not None:
    lawn_mask &= ~worm_spline.lab_frame_mask(center_tck, width_tck, timepoint_image.shape).astype('bool')
```

**Comment by drew-sinha** [comment]
Suggestions taken. I'll open up the PR for the measurements class shortly. I like your suggestion about using skimage's find_contours to vectorize the lawn image; long-term that might be more sustainable for making lawn measurements (+ is a little bit more in line with the prototypical workflow for the poses where one similarly vectorizes the masks as splines). That goes on the TODO list.

________________________________

File: elegant/segment_images.py [hunk 7]
```diff
@@ -7,11 +7,18 @@
 import subprocess
 import tempfile
 
+from scipy.ndimage import filters
+from scipy.ndimage import morphology
```
Comment (zpincus) [inline elegant/segment_images.py]: just do `from scipy import ndimage`: everything in the `filters` and `morphology` namespace is also just in the top-level `ndimage` as well.

File: elegant/segment_images.py [hunk 8]
```diff
@@ -140,6 +147,47 @@ def annotate_poses_from_masks(positions, mask_root, annotations, overwrite_exist
                 current_annotation[annotation] = pose
             current_annotation[original_annotation] = pose
 
+def gmm_lawn_maker(image, optocoupler, return_model=False):
```
Comment (zpincus) [inline elegant/segment_images.py]: *super minor* style point: names should simply describe what a function does, not how it does it. So maybe just call the function `find_lawn` or something (the GMM stuff is an implementation detail, not a key part of the external interface).

File: elegant/segment_images.py [hunk 9]
```diff
@@ -140,6 +147,47 @@ def annotate_poses_from_masks(positions, mask_root, annotations, overwrite_exist
...
+
+    # Calculate boundary point for label classification as intensity threshold
+    gmm_support = numpy.linspace(0,2**16-1,2**16)
+
+    labels = gmm.predict(numpy.reshape(gmm_support, (-1,1)))
+    thr = numpy.argmax(numpy.abs(numpy.diff(labels)))
+    lawn_mask = (filtered_image < thr) & vignette_mask
+
+    lawn_mask = morphology.binary_erosion(lawn_mask, iterations=10)
+    lawn_mask = zpl_mask.get_largest_object(lawn_mask)
```
Comment (zpincus) [inline elegant/segment_images.py:180]: should this maybe be the last function run? The dilation could join two objects, which I imagine you might want joined, and then you grab the largest contiguous object as a last step? Or is it actually better to grab a single object at the beginning?

Actually, I'm a little confused at what's going on. Usually you erode and then dilate to get rid of small objects, but you're already grabbing just the largest object. Or is the erosion and dilation to smooth the final object?

**Comment by drew-sinha** [comment]
Everything should be resolved.
